pallet/src/lib.rs

```diff
@@ -95,44 +73,43 @@ pub mod pallet {
     }
 
     impl<T: Config> Pallet<T> {
-        pub(crate) fn initialize_authorities(authorities: &[T::AuthorityId]) {
+        pub(crate) fn initialize_session_storage(authorities: Vec<T::AuthorityId>) {
             if !authorities.is_empty() {
-                assert!(
-                    <Authorities<T>>::get().is_empty(),
-                    "Authorities are already initialized!"
-                );
-                <Authorities<T>>::put(authorities);
+                if let Some(session) = <CurrentSession<T>>::get() {
+                    assert!(
+                        session.authorities.is_empty(),
+                        "Session is already initialized!"
+                    );
+                }
+                <CurrentSession<T>>::put(Session {
+                    session_id: 0,
+                    authorities: authorities.to_vec(),
+                    start_h: <frame_system::Pallet<T>>::block_number(),
+                    stop_h: <frame_system::Pallet<T>>::block_number() + T::Period::get().into(),
+                });
             }
-
-            <SessionInfo<T>>::put(SessionChange {
-                session_id: 0,
-                authorities_changed: true,
-                created_at: <frame_system::Pallet<T>>::block_number(),
-                next_authorities: authorities.to_vec(),
-            })
         }
 
-        pub(crate) fn update_authorities(authorities: &[T::AuthorityId]) {
-            <Authorities<T>>::put(authorities);
-        }
-
-        pub(crate) fn new_session(changed: bool, authorities: Vec<T::AuthorityId>) {
-            if let Some(old_session) = <SessionInfo<T>>::get() {
+        pub(crate) fn update_session_storage(
+            authorities: Vec<T::AuthorityId>,
+            queued_authorities: Vec<T::AuthorityId>,
+        ) {
+            if let Some(old_session) = <CurrentSession<T>>::get() {
                 let current_block = <frame_system::Pallet<T>>::block_number();
-
-                <SessionInfo<T>>::put(SessionChange {
+                <CurrentSession<T>>::put(Session {
                     session_id: old_session.session_id + 1,
-                    authorities_changed: changed,
-                    created_at: current_block,
-                    next_authorities: authorities,
+                    authorities,
+                    start_h: current_block,
+                    stop_h: current_block + T::Period::get(),
+                });
+                <NextSession<T>>::put(Session {
+                    session_id: old_session.session_id + 2,
+                    authorities: queued_authorities,
+                    start_h: current_block + T::Period::get(),
```

Comment:
Shouldn't start_h = stop_h + 1 of previous session?

Reply:
It should, thanks for noticing :)